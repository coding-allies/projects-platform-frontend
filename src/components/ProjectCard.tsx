import React, { FC } from "react";
import { Project, ExperienceLevelsTypes } from "../types";
import "../style/components/ProjectCard.css";

const getTags = (tagList: Array<string>) => {
  if (tagList.length > 5) {
    tagList = [...tagList.slice(0, 5)];
  }
  return tagList.map((tag) => {
    return (
      <div className={`tag`} key={tag}>
        {tag}
      </div>
    );
  });
};

const getContributors = (contributorList: Array<string>) => {
  if (contributorList.length > 6) {
    contributorList = [...contributorList.slice(0, 6), "..."];
  }

  return contributorList.map((contributor, i) => (
    <div className="card-contributor-icon" key={i}>
      {contributor[0].toUpperCase()}
    </div>
  ));
};

type Props = {
  data: Project;
  loginLink?: any;
};

const renderButtons = (project: Project, loginLink: any) => {
  if (!!project.lead.email) {
    return (
      <div className="card-buttons">
        <a
          href={project.github_url}
          className="button-link"
          target={project.github_url}
          rel="noopener noreferrer"
        >
          View on Github
        </a>
        <a
          href={`mailto:${project.lead.email}?subject=Request to join ${project.name}`}
          className="button-link"
        >
          Request to Join
        </a>
      </div>
    );
  } else {
    return (
      <div className="card-buttons">
        <button onClick={loginLink} className="button-link">
          View on Github
        </button>
        <button onClick={loginLink} className="button-link">
          Request to Join
        </button>
      </div>
    );
  }
};

const ProjectCard: FC<Props> = ({ data, loginLink }) => {
  const project = { ...data };
  const experienceLevel = ExperienceLevelsTypes[project.lead.experience];

  return (
    <article className="card-wrapper">
      <h2>{project.name}</h2>

      <div className="card-description">
        <p className="card-description-label">Project Description:</p>
        <p>{project.description}</p>
      </div>

      <p className="card-description">{project.description}</p>

      <div className="card-tech-stack">
        <p className="card-tech-stack-label">Tech stack: </p>
        <div className="card-tags">{getTags(project.tags)}</div>
      </div>

      <div className="card-looking-for">
        <p className="card-looking-for-label">Looking For:</p>
        <p>{project.looking_for}</p>
      </div>

      <div className="card-lead">
        <p className="card-lead-label">Project Lead:</p>
        <p>{project.lead.name}</p>
        <p>{project.lead.position}</p>
        <p>{experienceLevel}</p>
      </div>

      <div className="card-contributors">
        <p className="card-contributor-label">
          Contributors: {project.contributors.length}
        </p>
        <div className="card-contributor-avatars">
          {getContributors(project.contributors)}
        </div>
      </div>

      {renderButtons(project, loginLink)}
    </article>
  );
};

export default ProjectCard;
