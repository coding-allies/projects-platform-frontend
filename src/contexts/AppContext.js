import React, { createContext, Component } from "react";

export const AppContext = createContext();

const mockUser = {
  name: "Someone Awesome"
};

const mockProjects = [
  {
    projectName: "Project Name here",
    projectId: 18237,
    projectLead: {
      name: "Rey Dekker",
      yearsOfExperience: 5,
      currentPosition: "Facebook"
    },
    description:
      "This project aims to achieve...Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
    numberOfContributors: 3,
    contributors: ["NS", "MM", "NN", "IJ", "FR", "HJ", "WS"],
    tags: [
      "JS",
      "ReactJS",
      "Typescript",
      "TypescriptTypescript",
      "CSS",
      "Objective C"
    ]
  },
  {
    projectName: "Project 2 here context",
    projectId: 34534,
    projectLead: {
      name: "Jane Doe",
      yearsOfExperience: 1,
      currentPosition: "ABC Comp"
    },
    description:
      "Some short description - Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
    numberOfContributors: 1,
    contributors: ["JD"],
    tags: ["JS", "ReactJS"]
  }
];

export class AppContextProvider extends Component {
  state = {
    user: mockUser,
    projects: mockProjects
  };

  render() {
    return (
      <AppContext.Provider value={{ ...this.state }}>
        {this.props.children}
      </AppContext.Provider>
    );
  }
}
