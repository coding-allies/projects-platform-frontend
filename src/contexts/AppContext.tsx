import * as React from "react";
import { createContext, Component } from "react";
import { User, Project } from "../types";

export const AppContext = createContext({});

const mockUser: User = {
  name: "Someone Awesome"
};

const mockProjects: Array<Project> = [
  {
    projectName: "Project Name here",
    projectId: 18237,
    projectLead: {
      name: "Rey Dekker",
      experience: "Experienced (3+ years of experience)",
      company: "Facebook"
    },
    description:
      "This project aims to achieve...Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
    contributors: ["NS", "MM", "NN", "IJ", "FR", "HJ", "WS"],
    tags: [
      "JS",
      "ReactJS",
      "Typescript",
      "TypescriptTypescript",
      "CSS",
      "Objective C"
    ]
  },
  {
    projectName: "Project 2 here context",
    projectId: 34534,
    projectLead: {
      name: "Jane Doe",
      experience: "Learner (0+ years of experience)",
      company: "ABC Comp"
    },
    description:
      "Some short description - Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
    contributors: ["JD"],
    tags: ["JS", "ReactJS"]
  },
  {
    projectName: "Project 2 here context",
    projectId: 34534,
    projectLead: {
      name: "Jane Doe",
      experience: "Learner (0+ years of experience)",
      company: "ABC Comp"
    },
    description:
      "Some short description - Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
    contributors: ["JD"],
    tags: ["JS", "ReactJS"]
  },
  {
    projectName: "Project 2 here context",
    projectId: 34534,
    projectLead: {
      name: "Jane Doe",
      experience: "Learner (0+ years of experience)",
      company: "ABC Comp"
    },
    description:
      "Some short description - Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
    contributors: ["JD"],
    tags: ["JS", "ReactJS"]
  }
];

export class AppContextProvider extends Component<any, any> {
  state = {
    user: mockUser,
    projects: mockProjects
  };

  render() {
    return (
      <AppContext.Provider value={{ ...this.state }}>
        {this.props.children}
      </AppContext.Provider>
    );
  }
}
