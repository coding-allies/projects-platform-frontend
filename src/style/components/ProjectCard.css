.card-wrapper {
  padding: 24px;
  border: 2px solid #393939;
  display: flex;
  flex-direction: column;
  background-color: #FFF;
}

.card-wrapper h2 {
  margin: 0 0 16px;
  font-weight: 600;
  font-size: 1.4em;
  border-bottom: 3px solid turquoise;
}

.card-wrapper > * {
  padding-bottom: 16px;
}

.card-description {
  line-height: 21px;
  font-weight: 400;
}

.card-lead > * {
  font-size: 16px;
  line-height: 21px;
}

.card-tags {
  min-height: 68px;
  display: flex;
  flex-flow: wrap;
}

.tag {
  flex-shrink: 0;
  font-size: 14px;
  padding: 4px 10px 6px;
  margin: 0 8px 8px 0;
  border-radius: 1em;
  background-color: #c1fff8;
  max-height: 16px;
}

.card-contributors {
  padding-bottom: 16px;
}

.card-contributor-avatars {
  display: grid;
  grid-template-columns: repeat(auto-fit, 40px);
  grid-auto-rows: 40px 40px;
  grid-gap: 6px;
}

.card-description {
  min-height: 100px;
}

.card-looking-for {
  min-height: 130px;
}

.card-lead {
  min-height: 120px;
}

.card-description-label,
.card-looking-for-label,
.card-lead-label,
.card-contributor-label,
.card-tech-stack-label {
  text-transform: uppercase;
  font-weight: 600;
  letter-spacing: .6px;
  justify-self: left;
  font-size: 14px;
  padding-bottom: 6px;
  margin-bottom: 8px;
  border-bottom: 1px dotted #d1d1d3;

}

.card-contributor-icon {
  background-color: #393939;
  border-radius: 20px;
  text-align: center;
  padding-top: 10px;
  font-weight: 700;
  color: #FFF;
}

.card-buttons {
  display: grid;
  text-align: center;
  grid-template-columns: 1fr 1fr;
  grid-gap: 10px;
  padding-bottom: 0;
}
