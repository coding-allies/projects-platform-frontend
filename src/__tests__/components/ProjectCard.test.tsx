import * as React from "react";
import renderer from "react-test-renderer";
import ProjectCard from "../../components/ProjectCard";
import { mockProjects } from "../../contexts/AppContext";

it("renders ProjectCard correctly", () => {
  const data = mockProjects[0];
  const component = renderer
    .create(<ProjectCard data={data} />)
    .toJSON();
  expect(component).toMatchSnapshot();
});
