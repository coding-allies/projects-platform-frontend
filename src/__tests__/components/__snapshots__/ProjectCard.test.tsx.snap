// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ProjectCard correctly 1`] = `
<article
  className="card-wrapper"
>
  <h2>
    Project Name here
  </h2>
  <div
    className="card-lead"
  >
    <h3>
      Project Lead: 
      Ada Lovelace
    </h3>
    <p>
      Currently: 
      Self-employed
    </p>
    <p>
      Experienced (3+ years of experience)
    </p>
  </div>
  <p
    className="card-description"
  >
    This project aims to achieve...Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
  </p>
  <div
    className="card-looking-for"
  >
    <p
      className="card-looking-for-label"
    >
      Looking For:
    </p>
    <p>
      I could really use some help with enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
    </p>
  </div>
  <div
    className="card-contributors"
  >
    <p
      className="card-contributor-label"
    >
      Contributors: 
      7
    </p>
    <div
      className="card-contributor-avatars"
    >
      <div
        className="card-contributor-icon"
      >
        NS
      </div>
      <div
        className="card-contributor-icon"
      >
        MM
      </div>
      <div
        className="card-contributor-icon"
      >
        NN
      </div>
      <div
        className="card-contributor-icon"
      >
        IJ
      </div>
      <div
        className="card-contributor-icon"
      >
        FR
      </div>
      <div
        className="card-contributor-icon"
      >
        HJ
      </div>
      <div
        className="card-contributor-icon"
      >
        WS
      </div>
    </div>
  </div>
  <div
    className="card-tech-stack"
  >
    <p
      className="card-tech-stack-label"
    >
      Tech stack: 
    </p>
    <div
      className="card-tags"
    >
      <div
        className="tag"
      >
        JS
      </div>
      <div
        className="tag"
      >
        ReactJS
      </div>
      <div
        className="tag"
      >
        Typescript
      </div>
      <div
        className="tag"
      >
        TypescriptTypescript
      </div>
      <div
        className="tag"
      >
        CSS
      </div>
      <div
        className="tag"
      >
        Objective C
      </div>
    </div>
  </div>
  <div
    className="card-buttons"
  >
    <button
      className="card-button"
    >
      Github
    </button>
    <button
      className="card-button"
    >
      Favorite
    </button>
    <button
      className="card-button"
    >
      Join
    </button>
  </div>
</article>
`;
